_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+}
+```
+where the preceding `ext.kotlin_version = '1.0.+'` just selects the newest version of Kotlin from
+1.0 series.
+
+Finally, under `dependencies` section we need to state that Kotlin _stdlib_ has to be compiled.
+And as we are going to use Kotlin especially for testing purposes, it needs to be compiled for
+tests as well.
+
+__One note about tests and their location in the source tree.__
+
+As unit and UI tests are usually stored in `src/test/java` and `src/androidTest/java`
+respectively, we might want to maintain sanity in our codebase and keep Kotlin tests
+in dedicated directories like `src/test/kotlin` and `src/androidTest/kotlin`. As these
+are directories are not detected as Android test locations by default (*yet?*), we
+need to make Gradle and Android Studio aware of them. This can be done in the `sourceSets`
+section with
+
+```groovy
+android {
+    ...
+    sourceSets {
+        test.java.srcDirs += 'src/test/kotlin'
+        androidTest.java.srcDirs += 'src/androidTest/kotlin'
+    }
+}
+```
+
+## Our example app
+
+The [application](https://github.com/GreenhouseCI/kotlin-tests) created for
+the purposes of this blog post is of course yet another stunning shopping list app. As expected,
+it is possible to add some items to the list and remove them from there as you wish. To add
+some spice into the mix and make things a little bit more __testable__, every item that is
+inserted to the list passes through a formatter that removes all excess whitespace from the
+input string. Critical business logic in the form of our formatter that we use to clear up
+the strings lives in
+[`Formatter.java`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/main/java/com/greenhouseci/kotlin_tests/kotlintests/Formatter.java)
+and looks like
+
+```java
+public class Formatter {
+
+    public String stripMiddle(String s) {
+        return s.replaceAll("\\s+", " ");
+    }
+
+    public String stripLeft(String s) {
+        return s.replaceAll("^\\s+", "");
+    }
+
+    public String stripRight(String s) {
+        return s.replaceAll("\\s+$", "");
+    }
+
+    public String strip(String s) {
+        return stripLeft(stripRight(stripMiddle(s)));
+    }
+}
+```
+
+Bridge connecting user interactions with formatter resides conveniently in
+[`MainActivityFragment`](https://github.com/GreenhouseCI/kotlin-tests/blob/master/app/src/main/java/com/greenhouseci/kotlin_tests/kotlintests/MainActivityFragment.java#L38).
+
+All in all, our app stands in it's full glory as depicted below:
+<div style="text-align:center; padding: 20px 0;"><img src="/assets/kotlin-app-screenshot.png" alt="Screenshot"/></div>
+
+Evidently we want to test out that whenever item description `EditText` is filled with anything
+meaningful and "Add item" button is pressed, the newly added item is populated to the list.
+Furthermore, it is of utmost importance to ensure that all the items that are added to the list
```

Comment:
Can get rid of a bunch of small words: "to ensure that all ~~the~~ items ~~that are~~ added to the list"